File: 1516_2.py
from sys import stdin

input = stdin.readline


def solve():
    N = int(input())
    time = [0] * (N + 1)
    out_degree = [[] for _ in range(N + 1)]
    dp = [-1] * (N + 1)
    for i in range(1, N + 1):
        t, *need, _ = map(int, input().split())
        time[i] = t
        out_degree[i] = need
        if not need:
            dp[i] = t

    def dfs(node):
        if dp[node] != -1:
            return dp[node]

        dp[node] = max(dfs(nxt) for nxt in out_degree[node]) + time[node]
        return dp[node]

    ans = [0] * (N + 1)
    for i in range(1, N + 1):
        ans[i] = dfs(i)

    print(*ans[1:], sep="\n")


if __name__ == '__main__':
    solve()

File: 1516_2.js
const fs = require('fs');

function solve() {
	const input = fs.readFileSync('/dev/stdin').toString().split('\n');
	let inputIdx = 0;
	const N = +input[inputIdx++];
	const outDegree = [];
	const time = [];
	const ans = [];
	outDegree.length = N + 1;
	time.length = N + 1;
	ans.length = N + 1;
	for (let i = 1; i <= N; ans[i++] = -1);

	for (let i = 1; i <= N; i++) {
		const [t, ...need] = input[inputIdx++].split(' ').map((v) => +v);
		time[i] = t;
		outDegree[i] = need.slice(0, -1);
		if (outDegree[i].length === 0) ans[i] = t;
	}

	const dfs = (node) => {
		if (ans[node] !== -1) return ans[node];
		let max = -1;
		for (const nxt of outDegree[node]) {
			const tmp = dfs(nxt);
			max < tmp && (max = tmp);
		}
		return (ans[node] = max + time[node]);
	};

	for (let i = 1; i <= N; i++) {
		console.log(dfs(i));
	}
}

solve();
